alacritty/src/display/hint.rs

```diff
@@ -70,21 +73,57 @@ impl HintState {
             None => return,
         };
 
-        // Find visible matches.
-        self.matches.0 = hint.regex.with_compiled(|regex| {
-            let mut matches = RegexMatches::new(term, regex);
-
-            // Apply post-processing and search for sub-matches if necessary.
-            if hint.post_processing {
-                matches
-                    .drain(..)
-                    .map(|rm| HintPostProcessor::new(term, regex, rm).collect::<Vec<_>>())
-                    .flatten()
-                    .collect()
-            } else {
-                matches.0
-            }
-        });
+        let mut hint_matches: Vec<Match> = Vec::new();
+
+        // Find hyperlinks.
+        if hint.hyperlink {
+            // FIXME: These code dups with `RegexMatches::new`.
+            let viewport_start = Line(-(term.grid().display_offset() as i32));
+            let viewport_end = viewport_start + term.bottommost_line();
+
+            // Compute start of the first and end of the last line.
+            let mut start = term.line_search_left(Point::new(viewport_start, Column(0)));
+            let mut end = term.line_search_right(Point::new(viewport_end, term.last_column()));
+
+            // Set upper bound on search before/after the viewport to prevent excessive blocking.
+            start.line = max(start.line, viewport_start - MAX_SEARCH_LINES);
+            end.line = min(end.line, viewport_end + MAX_SEARCH_LINES);
+
+            let mut visited = HashSet::new();
+
+            traverse_hyperlinks(term, start, end, |hyperlink, bounds| {
+                // Only make the first part of a hyperlink is navigate-able, if there are many.
+                if visited.insert(hyperlink) {
+                    hint_matches.push(bounds);
+                }
+            });
+        }
+
+        // Find regex matches.
+        if let Some(regex) = &hint.regex {
+            regex.with_compiled(|regex| {
+                let matches = RegexMatches::new(term, regex);
+
+                // Apply post-processing and search for sub-matches if necessary.
+                if hint.post_processing {
+                    hint_matches.extend(
+                        matches
+                            .0
+                            .into_iter()
+                            .map(|rm| HintPostProcessor::new(term, regex, rm).collect::<Vec<_>>())
+                            .flatten(),
+                    );
+                } else {
+                    hint_matches.extend(matches.0);
+                }
+            });
+        }
+
+        // Sort and dedup ranges. Currently overlapped but not exactly same ranges are kept.
+        hint_matches.sort_by_key(|bounds| (*bounds.start(), Reverse(*bounds.end())));
+        hint_matches.dedup_by_key(|bounds| *bounds.start());
```

Comment:
I just use the most simple way -- sort and dedup. I'm not sure what's the better approach here when part of link text is URI, or part of URI-like text is a link. I think this mostly happen when a hyperlink is wrapped in tmux or something. In this case, the first part will be both regex-match and hyperlink-match, and we should prefer the latter, since regex-match get the wrong result. 
Complex overlapping case is very rare to happen in real world.